druid/src/lens.rs

```diff
@@ -141,3 +142,218 @@ where
         }
     }
 }
+
+/// Lens accessing a member of some type using accessor functions
+///
+/// See also the `lens` macro.
+///
+/// ```
+/// let lens = druid::lens::Field::new(|x: &Vec<u32>| &x[42], |x| &mut x[42]);
+/// ```
+pub struct Field<Get, GetMut> {
+    get: Get,
+    get_mut: GetMut,
+}
+
+impl<Get, GetMut> Field<Get, GetMut> {
+    /// Construct a lens from a pair of getter functions
+    pub fn new<T: ?Sized, U: ?Sized>(get: Get, get_mut: GetMut) -> Self
+    where
+        Get: Fn(&T) -> &U,
+        GetMut: Fn(&mut T) -> &mut U,
+    {
+        Self { get, get_mut }
+    }
+}
+
+impl<T, U, Get, GetMut> Lens<T, U> for Field<Get, GetMut>
+where
+    T: ?Sized,
+    U: ?Sized,
+    Get: Fn(&T) -> &U,
+    GetMut: Fn(&mut T) -> &mut U,
+{
+    fn with<V, F: FnOnce(&U) -> V>(&self, data: &T, f: F) -> V {
+        f((self.get)(data))
+    }
+
+    fn with_mut<V, F: FnOnce(&mut U) -> V>(&self, data: &mut T, f: F) -> V {
+        f((self.get_mut)(data))
+    }
+}
+
+/// Construct a lens accessing a type's field
+///
+/// This is a convenience macro for constructing `Field` lenses for fields or indexable elements.
+///
+/// ```
+/// struct Foo { x: u32 }
+/// let lens = druid::lens!(Foo, x);
+/// let lens = druid::lens!((u32, bool), 1);
+/// let lens = druid::lens!([u8], [4]);
+/// ```
+#[macro_export]
+macro_rules! lens {
+    ($ty:ty, [$index:expr]) => {
+        $crate::lens::Field::new::<$ty, _>(|x| &x[$index], |x| &mut x[$index])
+    };
+    ($ty:ty, $field:tt) => {
+        $crate::lens::Field::new::<$ty, _>(|x| &x.$field, |x| &mut x.$field)
+    };
+}
+
+/// `Lens` composed of two lenses joined together
```

Comment:
This is definitely one of the most important combinators, thanks! I think I'm good with "Then", but was also thinking about names more reminiscent of functional composition.

Reply:
I'm open to suggestions! This has some precedent in std, with [`Option::and_then`](https://doc.rust-lang.org/core/option/enum.Option.html#method.and_then) and moreso [`FutureExt::then`](https://docs.rs/futures/0.3.1/futures/future/trait.FutureExt.html#method.then).